### ISSUE TITLE
A shadowed @DefaultQualifier is lost for deeper subpackages when the shadowing default sets applyToSubpackages=false

### ISSUE THREAD

**Issue by wmdietl** [eisop/checker-framework#2037]
A `@DefaultQualifier` on a package is lost for deeper subpackages when an intervening package shadows it *and* sets `applyToSubpackages = false`. The intervening package's own default correctly stops at that package, but the enclosing one, which nothing shadows there, does not resume.

### Reproduction

```java
// A/package-info.java
@DefaultQualifier(value = Nullable.class, locations = TypeUseLocation.FIELD)
package A;

// A/B/package-info.java
@DefaultQualifier(value = Nullable.class, locations = TypeUseLocation.FIELD,
                  applyToSubpackages = false)
package A.B;
```

with the same class in each of `A`, `A.B` and `A.B.C`:

```java
public class T {
    Object f;

    void use() {
        f.toString();
    }
}
```

Running the Nullness Checker over all five files:

```
A/TA.java:8: error: [dereference.of.nullable] dereference of possibly-null reference f
A/B/TA_B.java:8: error: [dereference.of.nullable] dereference of possibly-null reference f
A/B/C/TA_B_C.java:4: error: [initialization.field.uninitialized] the default constructor does not initialize field f
```

In `A.B.C` the field is `@NonNull` -- the `initialization.field.uninitialized` error is the tell -- so neither default reached it.

Expected: `A`'s default applies in `A.B.C`. `A.B` limited only *its own* default to `A.B`; that should not remove `A`'s, which applies to subpackages.

### It is the shadowing, not the flag

Change `A.B`'s default to a different location and leave `applyToSubpackages = false`:

```java
@DefaultQualifier(value = Nullable.class, locations = TypeUseLocation.RETURN,
                  applyToSubpackages = false)
package A.B;
```

Now `A`'s FIELD default does reach `A.B.C`:

```
A/TA.java:8: error: [dereference.of.nullable] dereference of possibly-null reference f
A/B/TA_B.java:8: error: [dereference.of.nullable] dereference of possibly-null reference f
A/B/C/TA_B_C.java:8: error: [dereference.of.nullable] dereference of possibly-null reference f
```

So an enclosing package does normally reach through an inner `applyToSubpackages = false`. The loss happens only when the inner default shadows the outer one for the same (location, qualifier hierarchy).

### Mechanism

`QualifierDefaults.defaultsAt(Element)` (`QualifierDefaults.java:713`) computes each package's *effective* default set and caches it (`:774`). For a package it takes the parent's effective set and filters it to defaults with `applyToSubpackages` (`:735`), then merges the package's own defaults over it, dropping any parent default shadowed by a nearer one for the same (location, hierarchy) (`:756`).

At `A.B` that merge is right: `A.B`'s FIELD default shadows `A`'s, so `defaultsAt(A.B)` is `{d_A.B}`. But `d_A` has now been consumed, and when `A.B.C` filters its parent's effective set, `d_A.B` is dropped for `applyToSubpackages = false` and nothing remains. `d_A` is never reconsidered, even though nothing at `A.B.C` shadows it.

Shadowing is a statement about one scope; here it permanently removes a default from every deeper scope.

### Notes

Pre-existing, since #149 introduced `applyToSubpackages` and this filter; not related to #1996, which does not touch `QualifierDefaults`.

#1996 adds the same element to `@AnnotatedFor`, `@HasQualifierParameter` and `@ReportUse`, and those walk each enclosing package independently rather than collapsing to an effective set, so they do not have this behaviour: an enclosing package's annotation still applies at `A.B.C`. Whichever way this is resolved, it would be good for the two to agree.

A fix means not discarding a shadowed-but-still-propagating default -- either keeping it in the set with a marker, or walking ancestors independently and applying shadowing once at the end. `defaultsAt` is a hot, cached path, so this wants care rather than a quick patch.

It is also possible the current behaviour is intended -- "the nearest annotated package decides, and if it does not propagate then nothing does" is a coherent rule -- but nothing in the code or the manual says so, and it is not what the `applyToSubpackages` documentation describes.
